reflex-dom-core/src/Reflex/Dom/Xhr.hs

```diff
@@ -262,11 +263,16 @@ newXMLHttpRequestWithError req cb = do
     iforM_ (_xhrRequestConfig_headers c) $ xmlHttpRequestSetRequestHeader xhr
     maybe (return ()) (xmlHttpRequestSetResponseType xhr . fromResponseType) rt
     xmlHttpRequestSetWithCredentials xhr creds
+    -- Avoid handler being called more than once.
+    -- This is needed until jsaddle/ghcjs-dom: https://github.com/ghcjs/ghcjs-dom/issues/89 is fixed.
+    alreadyHandled <- liftIO $ newIORef False
     _ <- xmlHttpRequestOnreadystatechange xhr $ do
       readyState <- xmlHttpRequestGetReadyState xhr
       status <- xmlHttpRequestGetStatus xhr
       statusText <- xmlHttpRequestGetStatusText xhr
-      when (readyState == 4) $ do
+      handled <- liftIO $ atomicModifyIORef' alreadyHandled $ \handled ->
+        if readyState == 4 then (True, handled) else (handled, handled)
+      when (readyState == 4 && not handled) $ do
```

Comment:
Seems like it would make more sense to move the `handled` logic into `when` since both are checking for `readyState == 4` and it would avoid the `atomicModifyIORef'` when it's not relevant.

Reply:
Well, I just meant to make it thread safe just in case. Otherwise in a multi-threaded context, the handler could still be called twice.

Reply:
Moving the whole logic into the `when` would be thread safe as well but possibly be faster (fewer IORef operations) and not duplicate the `readyState == 4` logic.

Reply:
Ha right, now I get what you mean. Yep, that would be smarter :-)

Reply:
Ok fixed (untested), can't really remember why I thought in 2018 I would need that `atomicModifyIORef'` in the first place.

Reply:
Now I know why I thought the `atomicModifyIORef'` is needed, because it actually is ... *stupid me* My younger me, actually thought this through. Will fix in a second.